File: dartlang/lib/structures/stack.dart
class _Node<T> {
  final T value;
  _Node<T>? next;

  _Node(this.value);
}

class Stack<T> {
  _Node<T>? _head;

  int _size = 0;

  bool get isEmpty => _size == 0;

  int get size => _size;

  void push(T val) {
    final node = _Node<T>(val);
    if (_head != null) node.next = _head;
    _head = node;
    _size++;
  }

  T? pop() {
    final curr = _head;
    if (curr == null) return null;
    _head = _head?.next;
    _size--;
    return curr.value;
  }

  T? peek() => _head?.value;
}

File: php/structures/Stack.php
<?php

namespace Stack;

class Node
{
    public $next;
    function __construct(public $value) {}
}

//We don't do finds, if you need to find stuff just use an array,
//this is not a makeshift array
//
class Stack
{
    private int $size = 0;
    private ?Node $head = null;

    public function size()
    {
        return $this->size;
    }

    public function push(mixed $value)
    {
        $node = new Node($value);
        $node->next = $this->head;
        $this->head = $node;
        $this->size++;
    }

    public function pop()
    {
        if ($this->size == 0)  return null;
        $curr = $this->head->value;
        $this->head = $this->head->next;
        $this->size--;
        return $curr;
    }

    public function peek()
    {
        return $this->head->value;
    }
}
